gnowledge/gstudio
# MIS Update (Bug-fix for Course Announcement)

**Run following command after taking this pull request:**
1) Override GSTUDIO_TASK_TYPES variable in ```local_settings.py``` with followin value:
```GSTUDIO_TASK_TYPES = ["Bug", "Feature", "Support", "Student-Course Enrollment", "Student-Course Enrollment Approval", "UI Feature", "Other"]```
2) python manage.py filldb

<hr/>

**What to test?**
1) Try course announcement module (with different duration for given course-type)
2) Especially check whether you get to enroll students for these different duration announced courses
3) Try announcing courses using such duration for which there are no students registered 

<hr/>

**Bug-fix for Course-Announcement: Announcing course for different duration in same college**

1) views/course.py
  - Redundant code removed and placed at common place
    - For given colleges, fetching enrollment code, university to which they are affiliated to and assigned program officers
    - Previously code were written at different places which were querying in Triple
    - Now using college's node, fetching all these required information via ```enrollment_code``` (**attribute**), ```college_affiliated_to``` (**inverse-relation**), and ```has_officer_incharge``` (**relation**) respectively
  - **name** field modified for ```StudentCourseEnrollment``` GSystem and also of it's associated Tasks'
  - Adding **task-type** for respective task associated with ```StudentCourseEnrollment``` GSystem
  - Url link modified: Appending ```StudentCourseEnrollment``` GSystem's ObjectId to **mis_enroll** url

2) urls/mis.py
  - ```mis_enroll``` modified: ```app_set_instance_id``` appended that represents StudentEnrollment GSystem node's ObjectId

3) views/person.py
  - For given colleges, fetching enrollment code, university to which they are affiliated to
    - Previously querying in Triple
    - Now using college's node, fetching all these required information via ```enrollment_code``` (**attribute**) and ```college_affiliated_to``` (**inverse-relation**) respectively
  - **name** field modified for ```StudentCourseEnrollment``` GSystem and also of it's associated Tasks'
  - Adding **task-type** for respective task associated with ```StudentCourseEnrollment``` GSystem
  - Now populating announced course(s) that are open for enrollment from ```StudentCourseEnrollment``` GSystem node (using app_set_instance_id i.e. passed through ```mis_enroll``` url) directly via ```for_acourse``` (**relation**)

4) views/ajax_views.py
  - ```get_announced_courses_with_ctype()``` modified
    - Query modified that fetches announced course(s) for given college
      - Previously fetching from Triple; now making use of **inverse-relation** (```college_has_acourse```) from the college node itself
    - From the fetched announced course(s), keeping only those which satisfies following conditions:
      - which are active using **status** (```PUBLISHED```)
    - A special case also updated: Used only in ```new_create_batch.html```
      - Appending enrolled students count into announced course's node itself
      - Previously fetching from Triple; now making use of **inverse-relation** (```course_selected```) from the college node itself
  - ```get_anncourses_allstudents()``` modified
    - Query modified that fetches students for enrollment
      - Now fetching only those students whose registration date lies between given announced course's duration period
        - Hence, code added to fetch announced course's start_time and end_time, basically concerned with year
  - Preparing and populating drawer-widget only when any student exists for enrollment; otherwise simply a text is sent

5) templates/ndf/announced_course_create_edit.html
  - Year limit changed from current year to 2 years before current year for Course-announcement duration (```start_time```)

6) templates/ndf/student_enroll.html
  - Custom attributes declared for html element (acourse_list_name) removed (start_enroll_of_ac & end_enroll_of_ac)
  - Submit buttons (In Progress & Complete) moved into fieldset (fsSt)
    - Accordingly changed required javascript & css code
